src/braket/analog_hamiltonian_simulator/rydberg/validators/program.py

```diff
@@ -34,7 +39,75 @@ def local_detuning_pattern_has_the_same_length_as_atom_array_sites(cls, values):
+
+        # Extract the program and the fields
+        program = deepcopy(values)
+        del program["capabilities"]
+        program = Program.parse_obj(program)
+        driving_fields = program.hamiltonian.drivingFields
+        local_detuning = program.hamiltonian.localDetuning
+
+        # If no local detuning, we simply return the values
+        # because there are separate validators to validate
+        # the global driving fields in the program
+        if not len(local_detuning):
+            return values
+
+        detuning_times = [
+            local_detune.magnitude.time_series.times for local_detune in local_detuning
+        ]
+
+        # Merge the time points for different shifting terms and detuning term
+        time_points = sorted(list(set(sum(detuning_times, []))))
+        for driving_field in driving_fields:
+            time_points = sorted(list(set(time_points + driving_field.detuning.time_series.times)))
+
+        # Get the time-dependent functions for the detuning and shifts
+        _, detuning_coefs, shift_coefs = _get_coefs(program, time_points)
+
+        # Get the detuning pattern
+        detuning_patterns = [local_detune.magnitude.pattern for local_detune in local_detuning]
+
+        # For each time point, check that each atom has net detuning less than the threshold
+        for time_ind, time in enumerate(time_points):
+
+            # Get the contributions from all the global detunings
+            # (there could be multiple global driving fields) at the time point
+            values_global_detuning = sum(
+                [detuning_coef[time_ind] for detuning_coef in detuning_coefs]
+            )
+
+            for atom_index in range(len(detuning_patterns[0])):
+                # Get the contributions from local detuning at the time point
+                values_local_detuning = sum(
+                    [
+                        shift_coef[time_ind] * float(detuning_pattern[atom_index])
+                        for detuning_pattern, shift_coef in zip(detuning_patterns, shift_coefs)
+                    ]
+                )
+
+                # The net detuning is the sum of both the global and local detunings
+                detuning_to_check = np.real(values_local_detuning + values_global_detuning)
+
+                # Issue a warning if the absolute value of the net detuning is
+                # beyond MAX_NET_DETUNING
+                if abs(detuning_to_check) > capabilities.MAX_NET_DETUNING:
+                    warnings.warn(
+                        f"Atom {atom_index} has net detuning {detuning_to_check} rad/s "
+                        f"at time {time} seconds, which is outside the typical range "
+                        f"[{-capabilities.MAX_NET_DETUNING}, {capabilities.MAX_NET_DETUNING}]."
+                        f"Numerical instabilities may occur during simulation."
+                    )
+                    return values
```

Comment:
It still seems like this could be function of its own, but should be fine.

Reply:
Do you mean we have a function to check the validity of atom `i` at time point `j`, and then call that function in the for loop: `for atom_index ...`?